calico_rkt/calico_rkt.py

```diff
@@ -20,381 +20,474 @@
+
+        :return: None.
+        """
+        _log.info('Configuring pod %s' % self.container_id)
+
+        # Step 1: Assign an IP address using the given IPAM plugin.
+        assigned_ip = self._assign_ip()
+
+        # Step 2: Create the Calico endpoint object.
+        endpoint = self._create_endpoint(assigned_ip)
+
+        # Step 3: Provision the veth for this endpoint.
+        endpoint = self._provision_veth(endpoint)
+        
+        # Step 4: Provision / set profile on the created endpoint.
+        self._set_profile_on_endpoint(endpoint, self.network_name)
+
+        # Step 5: If all successful, print the IPAM plugin's output to stdout.
+        dump = json.dumps(self.ipam_result)
+        _log.info("Dumping info to container manager: %s", dump)
+        print(dump)
+
+        _log.info("Finished creating pod %s", self.container_id)
+    
+    def delete(self):
+        """Handles CNI_CMD_DELETE requests.
+
+        Remove this pod from Calico networking.
+
+        :return: None.
+        """
+        _log.info('Deleting pod %s' % self.container_id)
+
+        # Step 1: Get the Calico endpoint for this workload. If it does not
+        # exist, log a warning and exit successfully.
+        endpoint = self._get_endpoint()
+
+        # Step 2: The endpoint exists - remove any IP assignments.
+        self._release_ips()
+
+        # Step 3: Delete the veth interface for this endpoint.
+        _log.info("Removing veth for endpoint %s", endpoint.name)
+        netns.remove_veth(endpoint.name)
+
+        # Step 4: Delete the Calico endpoint.
+        self._remove_endpoint()
+
+        # Step 5: Delete any profiles for this endpoint
+        # TODO: This is racey and needs further thinking.
+        # self._remove_profile(self.network_name)
+
+        _log.info('Finished deleting pod %s' % self.container_id)
+
+    def _assign_ip(self):
+        """Assigns and returns an IPv4 address using the IPAM plugin
+        specified in the network config file.
+
+        :return: IPAddress - The assigned IP address.
+        """
+        # TODO - Handle errors thrown by IPAM plugin
+        self.ipam_result = self._call_ipam_plugin()
+        _log.debug("IPAM plugin result: %s", self.ipam_result)
 
-    _log.info('Deleting pod %s' % container_id)
+        try:
+            # Load the response and get the assigned IP address.
+            self.ipam_result = json.loads(self.ipam_result)
+        except ValueError:
+            _log.exception("Failed to parse IPAM response, exiting")
+            sys.exit(1)
 
-    # Remove the profile for the workload.
-    _container_remove(hostname=HOSTNAME,
-                      orchestrator_id=ORCHESTRATOR_ID,
-                      container_id=container_id)
+        # The request was successful.  Get the IP.
+        _log.info("IPAM result: %s", self.ipam_result)
+        return IPAddress(self.ipam_result["ip4"]["ip"])
+
+    def _release_ips(self, ip_list=[]):
```

Comment:
You shouldn't ever use a list object as the default value for a parameter - it will get re-used across multiple calls to this method.

Instead, you should do this:
```
def _release_ips(self, ip_list=None):
  if not ip_list:
    ip_list = []
```

As it stands, I don't think the `ip_list` argument is ever used, so it should probably just be removed.

Also, this should probably just be called `_release_ip` for symmetry with `_assign_ip`.